packages/core/src/tracing/instrumentation/messaging/rdkafka.js

```diff
@@ -0,0 +1,412 @@
+// with phase 2 it will no longer be configurable and will always use 'string'.
+let headerFormat = 'binary';
+
+let isActive = false;
+
+exports.init = function init(config) {
+  requireHook.onFileLoad(/\/node-rdkafka\/lib\/producer\.js/, instrumentProducer);
+  requireHook.onFileLoad(/\/node-rdkafka\/lib\/kafka-consumer-stream\.js/, instrumentConsumerAsStream);
+  requireHook.onModuleLoad('node-rdkafka', instrumentConsumer);
+  traceCorrelationEnabled = config.tracing.kafka.traceCorrelation;
+  headerFormat = config.tracing.kafka.headerFormat;
+};
+
+function instrumentProducer(ProducerClass) {
+  shimmer.wrap(ProducerClass.prototype, 'produce', shimProduce);
+}
+
+function instrumentConsumer(module) {
+  const className = 'KafkaConsumer';
+  const originalKafkaConsumer = module[className];
+
+  module[className] = function () {
+    const that = new originalKafkaConsumer(...arguments);
+    shimmer.wrap(that, 'emit', shimConsumerStreamEmit);
+    return that;
+  };
+
+  // We need to recover all static methods and attributes
+  // Currently, there is only one function: createReadStream
+  const keys = Object.keys(originalKafkaConsumer);
+
+  keys.forEach(key => {
+    module[className][key] = originalKafkaConsumer[key];
+  });
+}
+
+function instrumentConsumerAsStream(KafkaConsumerStream) {
+  shimmer.wrap(KafkaConsumerStream.prototype, 'emit', shimConsumerStreamEmit);
+}
+
+function shimProduce(originalProduce) {
+  return function () {
+    /**
+     * Arguments (ordered by their indexes):
+     * - topic: string,
+     * - partition: number,
+     * - message: Buffer,
+     * - key?: Kafka.MessageKey,
+     * - timestamp?: number,
+     * - opaque?: any,
+     * - headers?: Kafka.MessageHeader[]
+     */
+    const originalArgs = getFunctionArguments(arguments);
+    const message = originalArgs[2];
+
+    if (!isActive || !message) {
+      logger.debug('Will not instrument');
+      return originalProduce.apply(this, originalArgs);
+    }
+
+    return instrumentedProduce(this, originalProduce, originalArgs);
+  };
+}
+
+function shimConsumerStreamEmit(originalEmit) {
+  return function () {
+    /**
+     * Arguments (ordered by their indexes):
+     * - event name: string,
+     * - chunk: Object | any,
+     */
+    const originalArgs = getFunctionArguments(arguments);
+
+    if (!isActive) {
+      logger.debug('Will not instrument');
+      return originalEmit.apply(this, originalArgs);
+    }
+
+    return instrumentedEmit(this, originalEmit, originalArgs);
+  };
+}
+
+function instrumentedProduce(ctx, originalProduce, originalArgs) {
+  if (cls.tracingSuppressed()) {
+    const headers = addTraceLevelSuppression(originalArgs[6]);
+    originalArgs[6] = headers;
+    return originalProduce.apply(ctx, originalArgs);
+  }
+
+  const parentSpan = cls.getCurrentSpan();
+  if (!parentSpan || constants.isExitSpan(parentSpan)) {
```

Comment:
_General question_: What is the use case for parentSpan being already an exit span?
So there is an incoming message, which should be produced to a topic, but the parent span is already an exit span.
When does that happen?

Thanks!

Reply:
Essentially, the idea is to check if there isn't already a Kafka exit span being processed, or any kind of exit span for that matter.
As a rule of thumb, there should never be an ENTRY -> ENTRY span correlation, or EXIT -> EXIT. The expectation is to always have the alternated ENTRY -> EXIT -> ENTRY and so on.

If this is not clear enough, please let me know :)

Reply:
Thanks!